### DISCUSSION TITLE
Add new CSP support: NHN Cloud

### DISCUSSION THREAD

**Discussion by jihoon-seo** [cloud-barista/cb-tumblebug#1062] [General]
입력해야 하는 Credential fields: 
https://github.com/cloud-barista/cb-spider/blob/master/api-runtime/rest-runtime/test/connect-config/13.nhncloud-conn-config.sh 참고

```json
{
    "CredentialName":"nhncloud-credential01",
    "ProviderName":"NHNCLOUD",
    "KeyValueInfoList": [
        {"Key":"IdentityEndpoint", "Value":"https://api-identity.infrastructure.cloud.toast.com"},
        {"Key":"Username", "Value":"XXXXX@XXXXXXXXXXXXXXXX"},
        {"Key":"Password", "Value":"XXXXXXXXXXXXXXXXXX"},
        {"Key":"DomainName", "Value":"default"},
        {"Key":"TenantId", "Value":"XXXXXXXXXXXXXXXXX"}
]}
```
https://docs.toast.com/ko/Compute/Compute/ko/identity-api/
- `TenantId`: "웹 콘솔의 **Compute > Instance > 관리** 페이지의 **API 엔드포인트 설정** 대화 상자에서 확인합니다."
- `Password`: NHN Cloud 기본 인프라 서비스 API를 사용하려면 NHN Cloud 계정 비밀번호와는 별개로 API 비밀번호를 설정해야 합니다.
  1. **Compute > Instance > 관리** 페이지의 **API 엔드포인트 설정** 버튼을 클릭합니다.
  2. **API 엔드포인트 설정** 대화 상자 아래의 **API 비밀번호 설정**에 원하는 API 비밀번호를 지정합니다.

---

https://github.com/cloud-barista/nhncloud
> NHN Cloud infra가 논리적 네트워크 기반이지만, NHN Cloud에서 VPC 및 Subnet 생성/삭제 API는 지원하지 않고 조회 API만 제공하므로

현재 CB-Spider의 NHN Cloud 드라이버는 다음을 요구합니다.
- NHN Cloud 프로젝트의 기본 VPC가 존재하고, 그 이름이 'Default Network' 이어야 함
- 위 VPC의 기본 Subnet이 존재하고, 그 이름이 'Default Network' 이어야 함

위의 사항을 충족하려면, NHN Cloud 프로젝트를 생성하고, 기본적으로 존재하는 VPC 및 Subnet의 이름을 바꾸지 않으면 됩니다.

---

[Image, Spec, DiskType]
https://github.com/cloud-barista/nhncloud#-nhn-cloud-driver-사용시-참고-및-주의-사항

저의 경우, `rootDiskType` 을 빈칸으로 하고 VM 생성 요청을 할 때

- u2.c2m4 : 에러 발생
- m2.c4m8 : 에러 발생 없이 VM이 생성됩니다.

제가 VM 생성에 성공한 조합은 다음과 같습니다.

- South Korea (Pangyo) 리전
  - Image ID: 5396655e-166a-4875-80d2-ed8613aa054f
  - Spec ID: m2.c4m8
  - `rootDiskType`: 빈칸
- South Korea (Pyeongchon) 리전
  - Image ID: 71893505-923c-4982-b0eb-844eb520661c
  - Spec ID: m2.c4m8
  - `rootDiskType`: 빈칸
- Japan (Tokyo) 리전
  - Image ID: ca69fad9-92f3-4c9b-a6e4-7e8b0e4d3a58
  - Spec ID: m2.c4m8
  - `rootDiskType`: 빈칸

---

2022-04-08:
위의 사항들을 만족했을 때
TB 테스트 스크립트 수정 없이 ./create-all.sh & ./clean-all.sh 실행으로
자원 및 VM 생성/삭제가 되는 것을 확인했습니다.
(CC @itnpeople @powerkimhub @innodreamer @seokho-son)

**Comment by innodreamer** [comment]
@jihoon-seo (CC @powerkimhub @seokho-son)
NHN Cloud 현재 버전 기준으로 아래와 같이, u2.c2m4 type VM Spec 기준으로 RootDiskType을 빈칸("")으로 하고 VM 생성시 정상적으로 VM이 생성됩니다. (참고 : u2.c2m4 type VM Spec은 VM에 LocalDisk 로서 HDD, 50Gb만 지원)

CB-Spider API로 판교(KR1) region 기준으로 테스트시, 
O "RootDiskType": "", "RootDiskSize": "77" 로 지정해서 VM 생성 테스트했을때

[ 테스트 script ]
```
echo "## 4. VM: StartVM"
curl -sX POST http://localhost:1024/spider/vm -H 'Content-Type: application/json' -d '{ "ConnectionName": "'${CONN_CONFIG}'", "ReqInfo": { "Name": "nhn-vm-01", "ImageName": "'${IMAGE_NAME}'", "VPCName": "nhn-vpc-1", "SubnetName": "nhn-subnet-1", "SecurityGroupNames": [ "nhn-sg-02" ], "VMSpecName": "'${SPEC_NAME}'", "KeyPairName": "nhn-key-01", "RootDiskType": "", "RootDiskSize": "77" } }' |json_pp
```
[ 테스트 결과 ]
```
## 4. VM: StartVM
{
   "message" : "Invalid RootDiskSize!! Specified VMSpec [u2.c2m4] supports only [50](GB)!!"
}
```

O "RootDiskType": "", "RootDiskSize": "50" 으로 지정해서 VM 생성 테스트했을때

[ 테스트 script ]
```
echo "## 4. VM: StartVM"
curl -sX POST http://localhost:1024/spider/vm -H 'Content-Type: application/json' -d '{ "ConnectionName": "'${CONN_CONFIG}'", "ReqInfo": { "Name": "nhn-vm-01", "ImageName": "'${IMAGE_NAME}'", "VPCName": "nhn-vpc-1", "SubnetName": "nhn-subnet-1", "SecurityGroupNames": [ "nhn-sg-02" ], "VMSpecName": "'${SPEC_NAME}'", "KeyPairName": "nhn-key-01", "RootDiskType": "", "RootDiskSize": "50" } }' |json_pp
```

[ 테스트 결과 ]
```
## 4. VM: StartVM
{
   "VMBlockDisk" : "",
   "VpcIID" : {
      "SystemId" : "ae945890-0433-467e-9366-b9e3611e01f3",
      "NameId" : "nhn-vpc-1"
   },
   "IId" : {
      "SystemId" : "ceb40340-dea0-444d-9bcd-23ce74095057",
      "NameId" : "nhn-vm-01"
   },
   "StartTime" : "2022-05-23T10:41:58Z",
   "VMUserPasswd" : "N/A",
   "RootDiskSize" : "50",
   "PublicIP" : "125.6.37.155",
   "Region" : {
      "Region" : "KR1",
      "Zone" : "N/A"
   },
   "ImageIId" : {
      "NameId" : "5396655e-166a-4875-80d2-ed8613aa054f",
      "SystemId" : "5396655e-166a-4875-80d2-ed8613aa054f"
   },
   "PublicDNS" : "",
   "SecurityGroupIIds" : [
      {
         "SystemId" : "67167e2e-2390-48d6-8f27-78c9293b26f3",
         "NameId" : "nhn-sg-02"
      }
   ],
   "KeyPairIId" : {
      "NameId" : "nhn-key-01",
      "SystemId" : "nhn-key-01-c9584r9jcupvtimg81l0"
   },
   "NetworkInterface" : "a1184c28-7c30-4d2c-9af4-a82dfd606e45",
   "RootDiskType" : "VM_Local_Disk",
   "PrivateIP" : "192.168.0.35",
   "SubnetIID" : {
      "SystemId" : "fa6ddcee-9761-433a-af9e-6334b3a33f25",
      "NameId" : "nhn-subnet-1"
   },
   "PrivateDNS" : "",
   "SSHAccessPoint" : "125.6.37.155:22",
   "KeyValueList" : [
      {
         "Value" : "2",
         "Key" : "vCPU"
      },
      {
         "Value" : "4096",
         "Key" : "vRAM(GB)"
      }
   ],
   "VMUserId" : "cb-user",
   "VMBootDisk" : "",
   "RootDeviceName" : "",
   "VMSpecName" : "u2.c2m4"
}
```

**Reply by innodreamer** [reply]
@jihoon-seo (CC @seokho-son  @powerkimhub )

아래의 cb-spider issue에 link에 정리된 google docs 문서 내용중 NHN Cloud 부분 참고하세요.
https://github.com/cloud-barista/cb-spider/issues/598#issuecomment-1097610395

**Comment by jihoon-seo** [comment]
2022년 9월 26일 기준, 
CB-Spider 및 CB-Tumblebug 최신버전 (latest master/main) 에서 
CB-Tumblebug 테스트 스크립트 수정 없이
NHN Cloud Pangyo 1 (Region: `KR1` // Zone: `kr-pub-a`) 에 대해
VM 생성 및 NLB 생성/VM삭제/VM추가/삭제 까지 되는 것을 확인했습니다.

- CB-Spider: v0.6.9 + a ([`90f305e`](https://github.com/cloud-barista/cb-spider/commit/90f305eb21515cf1d377265c30f8e5757383d0d4))
- CB-Tumblebug: v0.6.2 + a (1dda901)

[참고]
9월 6일에 CB-Spider에서 NHN Cloud에 대해 AZ 정보를 추가해 주셨습니다. ([관련 PR](https://github.com/cloud-barista/cb-spider/pull/740))
이를 CB-Tumblebug 테스트 스크립트에도 반영했습니다.

**Comment by seokho-son** [comment]
./create-all.sh -f ../testSet-nhn.env -n nhn01
![image](https://user-images.githubusercontent.com/5966944/193758964-2a1b4d4a-37eb-40b3-b77a-697395e0274f.png)
![image](https://user-images.githubusercontent.com/5966944/193759045-d5e7473f-1d15-4fda-a61a-21d3d99de655.png)

./gen-sshKey.sh -f ../testSet-nhn.env -n nhn01
![image](https://user-images.githubusercontent.com/5966944/193759167-d7ecff48-304c-411e-a2ef-6c2f6b5c65ef.png)

**Comment by jihoon-seo** [comment]
customImage, dataDisk 기능 테스트 완료
- VM 스냅샷 -> customImage 생성
- customImage 삭제
- customImage로 VM 생성
- dataDisk 생성, 삭제
- VM에 dataDisk attach, detach

---

(테스트 flow)

[DataDisk가 붙어 있지 않은 VM의 snapshot]
OK
(TB customImage object 생성됨)
![image](https://user-images.githubusercontent.com/46767780/203892494-f8a1ad56-4911-4922-ab8a-3502bc78dd1a.png)

[customImage 삭제]
```
❯ ./delete-customimage.sh -n jhseo -c nhncloud -r 1
####################################################################
## 6. image: Unregister
####################################################################


- Unregister image in nhncloud-kr-pangyo1
{
  "message": "The customImage nhncloud-kr-pangyo1-jhseo has been deleted"
}
```

(삭제되었음을 NHN Cloud 콘솔에서도 확인함)

[dataDisk 생성]
OK

[생성한 dataDisk를 VM에 attach]
OK

[dataDisk가 attach 되어 있는 VM의 스냅샷 생성]
OK
(단, NHN Cloud는 "VM 스냅샷" 이라는 메뉴는 없으며, "커스텀 이미지" 가 생성될 뿐이고, 
생성된 커스텀 이미지에는 VM rootDisk의 내용이 포함되어 있으며, attach 되어 있던 dataDisk에 대한 정보나 흔적은 없음)
![image](https://user-images.githubusercontent.com/46767780/203910382-173ac0ee-52f4-4b04-8ed9-252344da87f3.png)

[(customImage로 VM을 실행하기 위해) TB image object 삭제]
OK

[생성했던 VM 삭제]
OK

[customImage로 VM 생성]
OK

[VM에 dataDisk attach & detach]
OK

**Comment by jihoon-seo** [comment]
(NLB 테스트)

[동일 MCIS, 동일 subGroup에 속하는 VM 2개 생성]
OK

[NLB 생성 시도 1]
(NHN Cloud 콘솔에서도 신규 NLB 항목이 (생성 중일 때) 보였다가 사라지고 아래 에러 리턴)

> ❯ ./create-NLB.sh -n jhseo -c nhncloud -r 1      
> ####################################################################
> ## 10. NLB: Create
> ####################################################################
> 
> {
>   "message": "{\"message\":\"Failed to Create Pool. [Failed to Create NHN Cloud Pool. [Expected HTTP response code [201 202] when accessing [POST https://kr1-api-network.infrastructure.cloud.toast.com/v2.0/lbaas/pools], but got 409 instead\\n{\\\"NeutronError\\\": {\\\"message\\\": \\\"Invalid state PENDING_UPDATE of loadbalancer resource 8736b880-a380-4457-9995-17037554d16d\\\", \\\"type\\\": \\\"StateInvalid\\\", \\\"detail\\\": \\\"\\\"}}]]\"}\n"
> }

[NLB 생성 시도 2]
(NHN Cloud 콘솔에서도 신규 NLB 항목이 (생성 중일 때) 보였다가 사라지고 아래 에러 리턴)

> ❯ ./create-NLB.sh -n jhseo -c nhncloud -r 1
> ####################################################################
> ## 10. NLB: Create
> ####################################################################
> 
> {
>   "message": "{\"message\":\"Failed to Create HealthMonitor. [Failed to Create HealthChecker. [Expected HTTP response code [201 202] when accessing [POST https://kr1-api-network.infrastructure.cloud.toast.com/v2.0/lbaas/healthmonitors], but got 409 instead\\n{\\\"NeutronError\\\": {\\\"message\\\": \\\"Invalid state PENDING_UPDATE of loadbalancer resource 10a0ac60-0cd1-4437-9a3a-00a160687f0a\\\", \\\"type\\\": \\\"StateInvalid\\\", \\\"detail\\\": \\\"\\\"}}]]\"}\n"
> }

[NLB 생성 시도 3]
(NHN Cloud 콘솔에서도 신규 NLB 항목이 (생성 중일 때) 보였다가 사라지고 아래 에러 리턴)

> ❯ ./create-NLB.sh -n jhseo -c nhncloud -r 1
> ####################################################################
> ## 10. NLB: Create
> ####################################################################
> 
> {
>   "message": "{\"message\":\"Failed to Create HealthMonitor. [Failed to Create HealthChecker. [Expected HTTP response code [201 202] when accessing [POST https://kr1-api-network.infrastructure.cloud.toast.com/v2.0/lbaas/healthmonitors], but got 409 instead\\n{\\\"NeutronError\\\": {\\\"message\\\": \\\"Invalid state PENDING_UPDATE of loadbalancer resource c31bf352-e4d9-4d5f-b0fa-8e445c67b543\\\", \\\"type\\\": \\\"StateInvalid\\\", \\\"detail\\\": \\\"\\\"}}]]\"}\n"
> }

(해당 repo에 이슈 오픈하였음)
